### ISSUE TITLE
dqn时序差分目标没加detach

### ISSUE THREAD

**Issue by weibin13** [boyu-ai/Hands-on-RL#112]
dqn时序差分目标不加detach吗，目标网络也累积梯度？

**Comment by nayouki** [comment]
我也想问了，明明PPO开始都加了。但DQN处理有点特殊，对目标网络的值有取最大值的操作，这个也显然不可导吧？但据说计算的时候还是会给最大值来源那个变量分配梯度。。。。我测试了两种算法结果，比较相近。但我还是倾向于加上detach

**Comment by nayouki** [comment]
又找到了个新说法，实际上由于目标网络中我们并没有给它分配优化器，因为每次step的时候DQN都只能更新自己的网络参数，所以是可加可不加？（而且从效果上看没啥区别）PPO中由于它没有目标网络机制，每次都用同一个网络，所以必须在old的情况下加上detach。到了DDPG和目标网络相关的target也没加detach。所以我认为是优化器和网络参数的绑定

**Comment by nayouki** [comment]
这也解释了为为什么DDPG中对actor更新时没有加detach防止更新到critic，因为后面只调用了actor优化器的step，不会干扰到critic
在代码书写上，PPO是最后一起调用两个优化器；而到了DDPG则改为有顺序的调用优化器，这也侧面说明了这样的安排是为了防止互相干扰

**Comment by NEMOlv** [comment]
> 又找到了个新说法，实际上由于目标网络中我们并没有给它分配优化器，因为每次step的时候DQN都只能更新自己的网络参数，所以是可加可不加？（而且从效果上看没啥区别）PPO中由于它没有目标网络机制，每次都用同一个网络，所以必须在old的情况下加上detach。到了DDPG和目标网络相关的target也没加detach。所以我认为是优化器和网络参数的绑定

你的理解是对的。需要将网络参数传入指定优化器，并且在loss反向传播之后执行优化器的step网络参数才最发生更新。如果目标网络连优化器都没有，是否加detach是无所谓的

**Comment by BruceZhang111** [comment]
> 又找到了个新说法，实际上由于目标网络中我们并没有给它分配优化器，因为每次step的时候DQN都只能更新自己的网络参数，所以是可加可不加？（而且从效果上看没啥区别）PPO中由于它没有目标网络机制，每次都用同一个网络，所以必须在old的情况下加上detach。到了DDPG和目标网络相关的target也没加detach。所以我认为是优化器和网络参数的绑定

我也实际测试并对比了加与不加 `.detach()` 的情况，效果如下：
1. 对训练结果的影响：由于在定义优化器时，仅将主网络（`q_net`）的参数传入了 Adam，目标网络（`target_q_net`）并未交由优化器管理。因此，即使在计算时向后传导了梯度，目标网络的参数权重也不会在 `optimizer.step()` 中被实际更新。所以加与不加 `.detach()` 在最终的训练表现和收敛曲线上几乎没有区别。
2. 对硬件资源的影响：尽管结果一致，但如果不加 `.detach()`，PyTorch 会在每次前向传播时将目标网络纳入计算图谱中，并在反向传播时计算并暂存不必要的梯度。在我自己的环境实测中，加入 `.detach()` 可以让训练时的 GPU 显存占用降低大约 10%。
